tasks/licenses.py

```diff
@@ -0,0 +1,95 @@
+import json
+import pathlib
+import time
+from contextlib import closing
+from io import StringIO
+
+import httpx
+
+LATEST_SPDX_GITHUB_RELEASE_URL = (
+    "https://api.github.com/repos/spdx/license-list-data/releases/latest"
+)
+LICENSES_URL = (
+    "https://raw.githubusercontent.com/spdx/license-list-data/v{}/json/licenses.json"
+)
+EXCEPTIONS_URL = (
+    "https://raw.githubusercontent.com/spdx/license-list-data/v{}/json/exceptions.json"
+)
+
+
+def download_data(url):
+    for _ in range(600):
+        try:
+            response = httpx.get(url)
+            response.raise_for_status()
+        except Exception:
+            time.sleep(1)
+            continue
+        else:
+            return json.loads(response.content.decode("utf-8"))
+
+    message = "Download failed"
+    raise ConnectionError(message)
+
+
+def main():
+    latest_version = download_data(LATEST_SPDX_GITHUB_RELEASE_URL)["tag_name"][1:]
+
+    licenses = {}
+    for license_data in download_data(LICENSES_URL.format(latest_version))["licenses"]:
+        license_id = license_data["licenseId"]
+        deprecated = license_data["isDeprecatedLicenseId"]
+        licenses[license_id.lower()] = {"id": license_id, "deprecated": deprecated}
+
+    exceptions = {}
+    for exception_data in download_data(EXCEPTIONS_URL.format(latest_version))[
+        "exceptions"
+    ]:
+        exception_id = exception_data["licenseExceptionId"]
+        deprecated = exception_data["isDeprecatedLicenseId"]
+        exceptions[exception_id.lower()] = {
+            "id": exception_id,
+            "deprecated": deprecated,
+        }
+
+    project_root = pathlib.Path(__file__).resolve().parent.parent
+    data_file = project_root / "src" / "packaging" / "licenses" / "spdx.py"
+
+    with closing(StringIO()) as file_contents:
+        file_contents.write(
+            f"""\
+from __future__ import annotations
+
+from typing import TypedDict
+
+class SPDXLicense(TypedDict):
+    id: str
+    deprecated: bool
+
+class SPDXException(TypedDict):
+    id: str
+    deprecated: bool
+
+
+VERSION = {latest_version!r}
+
+LICENSES: dict[str, SPDXLicense] = {{
+"""
```

Comment:
nit-ish: Could we indent this, so that it doesn't break the visual hierarchy that the indentation provides?

```suggestion
            textwrap.dedent(f"""\
                from __future__ import annotations

                from typing import TypedDict

                class SPDXLicense(TypedDict):
                    id: str
                    deprecated: bool

                class SPDXException(TypedDict):
                    id: str
                    deprecated: bool


                VERSION = {latest_version!r}

                LICENSES: dict[str, SPDXLicense] = {{
           """)
```

+ whatever the linters do.

Reply:
9cee38e